### ISSUE TITLE
Multiple servers support

### ISSUE THREAD

**Issue by maximyurevich** [yegappan/lsp#228]

**Comment by andlrc** [comment]
How would things like `LspGotoDefinition`, `LspPeekDefinition`, `LspPeekReferences` etc. work if two servers reports different locations?

How would things like `LspCodeAction` work? Should the actions be merged into one list? If so what happens with potential "duplicates"?

What happens with `LspFold` if the two servers reports different folds, `LspHighlight` if the server reports different positions, etc? Should the user have to specify which running server that should be used?

I think that there are many questions that should be answered before it makes sense to add support for multiple servers? Is there somewhere else that we can draw inspiration from?

**Comment by maximyurevich** [comment]
> How would things like `LspGotoDefinition`, `LspPeekDefinition`, `LspPeekReferences` etc. work if two servers reports different locations?
> 
> How would things like `LspCodeAction` work? Should the actions be merged into one list? If so what happens with potential "duplicates"?
> 
> What happens with `LspFold` if the two servers reports different folds, `LspHighlight` if the server reports different positions, etc? Should the user have to specify which running server that should be used?
> 
> I think that there are many questions that should be answered before it makes sense to add support for multiple servers. Is there somewhere else that we can draw inspiration from? I'm sure VSCode have this stuff figured out?

For inspiration https://github.com/prabirshrestha/vim-lsp
I want to add [efm-langserver](https://github.com/mattn/efm-langserver) as second LSP for linting and formatting

**Comment by andlrc** [comment]
> For inspiration https://github.com/prabirshrestha/vim-lsp

I not sure I understand how in works in the linked plugin, can you show a few examples for how the above mentioned concerns are addressed?

**Comment by yegappan** [comment]
I did a google search for the multiple language servers per file support and found the following relevant links:

https://github.com/emacs-lsp/lsp-mode/issues/424
https://github.com/Microsoft/vscode-extension-samples/tree/main/lsp-multi-server-sample
https://github.com/helix-editor/helix/pull/2507
https://neovim.discourse.group/t/multiple-parallel-langauge-servers/235
https://www.reddit.com/r/neovim/comments/jy08md/multiple_languageserver_choose_format/
https://neovim.discourse.group/t/automatically-choose-one-language-server-to-format-code-when-using-multiple-language-servers/800/7
https://github.com/natebosch/vim-lsc/issues/57

I think we should allow the user to configure which LSP server should provide a particular LSP functionality.
But for some functionality like diagnostics, we can allow them to use multiple servers and merge the results.

**Comment by Shane-XB-Qian** [comment]
simple way is two init cfg with two vim instances by user manually.
e.g deno with cfg_deno, tsserver with cfg_ts, in two vim instances, though same/one proj.
specially if someone created deno.json AND package.json both in one/same proj, that would be confused and complex to make it work via one/same lsp client in one vim instance.

**Comment by andlrc** [comment]
I started the work here: https://github.com/yegappan/lsp/pull/231 I think we might be able to merge the first iteration soon

**Comment by lithammer** [comment]
I think maybe a solution to some of these is to allow the user to set a priority on either a language server as a whole, or individual features of a language server.

Using the latter, you would be able to cherry-pick for example completion from server A and formatting from server B.

Using a priority instead of enable/disable means it would work better across different projects. One project might only have server B for example, and then you want completion from that instead of A.

I think this is partly how CoC solves it.

**Comment by maximyurevich** [comment]
> I started the work here: #231 I think we might be able to merge the first iteration soon

Hello, this pr doesn't work for me with ruff-lsp and pylsp

**Comment by Shane-XB-Qian** [comment]
> I think maybe a solution to some of this is to allow the user to set a priority on either a language server as a whole, or individual features of a language server.
> 
> Using the latter, you would be able to cherry-pick for example completion from server A and formatting from server B.
> 
> Using a priority instead of enable/disable means it would work better across different projects. One project might only have server B for example, and then you want completion from that instead of A.
> 
> I think this is partly how CoC solves it.

actually we had done the latter one, please check the `omnicompl` option in cfg which i impl it perhaps about 2 years ago.

you can chose one as the major lsp server, another one is to help it e.g omnicompl, which that feature is a feature also would not impact/conflict each others too much.
`toggle` way probably was the less-conflict way, e.g #233 though i did not check that issue detail but i guess it should be some others plugins conflicted with lsp function, and that plugs perhaps had `toggle` option of that feature but no `priority` such thing from it with this lsp client.

i did not check #231 detail much yet, but seems that changed code much at once, actually i am not feeling comfortable that, it is difficult to verify change if so.
meanwhile, i think we can just pick-up some less-conflict feature (NOT all features) to impl multi-servers support if preferred, e.g 'diag' which perhaps can be impl same way like omnicompl, vs/and some lsp server e.g efm-langserver only offerred 'diag' feature, then these two can be work together without impact.
vs if to `deno` and `tsserver`, to me, i perferred DO NOT let them work in same one vim instance together at same time, (unless deno or tsserver allowed user just pick some of its feature), i can simply switch lsp-cfg to archive that.

**Comment by Shane-XB-Qian** [comment]
@andlrc though #231 had been merged, but still not understand how it worked with 'rootSearch'

if set tsserver: rootsearch: ['.git/', 'package.json'], 
and/then deon: rootsearch: ['.git/', 'deno.json']

1, to a non git but had package.json ts proj, what's the lsp server and workspace it should be?
2, to a non git but had deno.json ts proj, what's the lsp server and workspace it should be?

i saw to a git project, it started both, but to a non-git project, i do not understand how it played...

**Comment by andlrc** [comment]
> @andlrc though #231 had been merged, but still not understand how it worked with 'rootSearch'
> 
> if set tsserver: rootsearch: ['.git/', 'package.json'], and/then deon: rootsearch: ['.git/', 'deno.json']
> 
> 1, to a non git but had package.json ts proj, what's the lsp server and workspace it should be? 2, to a non git but had deno.json ts proj, what's the lsp server and workspace it should be?
> 

This is not really related to #231 as the behaviour is almost the same as before support for multiple servers was added.

Previously the server with the longest rootSearch path was used, now all that share the same length rootSearch path will be used.

> i saw to a git project, it started both, but to a non-git project, i do not understand how it played...

If the `.git` directory cannot be found, or is found after one of the other patterns, then the containing directory is used for score logic:

Consider the following structure:

    /root
      |- folderA/
          |- package.json
          |- folderB
              |- deno.json
              
- If you open a file within `folderB` then `deno.json` is the first match and deon is used.
- If you open a file within `folderA` that are not within `folderB` then `package.json` is the first match and tsserver is used.
- If you open a file with `root` but not within `folderA` nor `folderB` then both servers are being used, as they score the same.

I previously purposed a `runIf` configuration that takes a funcref which should return a boolean indicating if a server should run or not, but it was opted to not do that yet. I still have the code for it, so if it makes sense, then we could implement that on top.

**Comment by Shane-XB-Qian** [comment]
to non git proj, looks it always started the firstly configured lsp server, regardless the rootseatch. I recall (at my morning) it behaved like that, but it should be a wrong logic, did not respect the rootseatch at all.

if they are same longest rootseatch matched (find each same ft lsp server itself longest first), then to concern which one configured firstly, and/next to concern which 'features' configured enabled.
I think this should be the reasonable logic.

--
shane.xb.qian

**Comment by andlrc** [comment]
> but it should be a wrong logic, did not respect the rootseatch at all.

`rooSearch` serves two purposes currently:

1. Find the root of the project, if not found then cwd or buffer dir is used.
2. Determinate what servers to run, based on score. The score is -1 for servers with a non matching rootSearch and 0 for servers without a `rootSearch`.

It might make sense to avoid 2. completely by adding another option. `runSearch` or something?

**Comment by Shane-XB-Qian** [comment]
I did not mean `avoid`, but the determination to the score seems incorrectly which caused server started (and/or init workspace config) unexpectedly.

but anyway I still felt allow such multiple server for same ft at same time same instance made code & logic complex which may take much time/cost to make it be right.

--
shane.xb.qian
________________________________

**Comment by maximyurevich** [comment]
Hello @andlrc, multiple servers starting working only after ```LspServer restart```. Is this normal behavior for now?

**Comment by andlrc** [comment]
> Hello @andlrc, multiple servers starting working only after `LspServer restart`. Is this normal behavior for now?

No, how have you configured your servers?

**Comment by maximyurevich** [comment]
> > Hello @andlrc, multiple servers starting working only after `LspServer restart`. Is this normal behavior for now?
> 
> No, how have you configured your servers?

" LSP

let g:mason_dir = $HOME.'/.local/share/nvim/mason/bin/'
```vim
let g:lspServers = [
	\  {
	\	    'name': 'typescript',
	\	    'filetype': ['javascript', 'typescript', 'javascriptreact', 'typescriptreact'],
	\	    'path': mason_dir.'typescript-language-server',
	\	    'args': ['--stdio'],
  \     'rootSearch': ['package.json']
	\   },
	\   {
	\	    'name': 'deno',
	\	    'filetype': ['javascript', 'typescript', 'javascriptreact', 'typescriptreact'],
	\	    'path': 'deno',
	\	    'args': ['lsp'],
  \     'initializationOptions': {
  \       'enable': v:true,
  \       'lint': v:true
  \     },
  \     'rootSearch': ['deno.json', 'deno.jsonc']
	\   },
...
	\   {
	\	    'name': 'efm-langserver',
	\	    'filetype': [
  \                   'vim',
  \                   'javascript', 
  \                   'typescript', 
  \                   'javascriptreact', 
  \                   'typescriptreact',
  \                   'python',
  \                   'ruby'
  \                 ],
	\	    'path': 'efm-langserver',
	\	    'args': [],
  \   },
	\ ]

augroup lsp_servers
  autocmd VimEnter * silent! call LspAddServer(lspServers)
augroup END

let lspOpts = {
  \   'noNewlineInCompletion': v:true,
  \   'hideDisabledCodeActions': v:true,
  \   'autoPopulateDiags': v:true,
  \   'usePopupInCodeAction': v:true,
  \   'autoHighlightDiags': v:true,
  \   'ultisnipsSupport': v:true,
  \   'snippetSupport': v:true,
  \   'outlineOnRight': v:true,
  \   'outlineWinSize': 40,
  \	  'customCompletionKinds': v:true,
  \	  'completionKinds': {
  \     'Text': '',
  \     'Method': '',
  \     'Function': '',
  \     'Constructor': '',
  \     'Field': 'ﰠ',
  \     'Variable': '',
  \     'Class': 'ﴯ',
  \     'Interface': '',
  \     'Module': '',
  \     'Property': 'ﰠ',
  \     'Unit': '塞',
  \     'Value': '',
  \     'Enum': '',
  \     'Keyword': '',
  \     'Snippet': '',
  \     'Color': '',
  \     'File': '',
  \     'Reference': '',
  \     'Folder': '',
  \     'EnumMember': '',
  \     'Constant': '',
  \     'Struct': 'פּ',
  \     'Event': '',
  \     'Operator': '',
  \     'TypeParameter': '',
  \	  }
  \ }

augroup lsp_options
  autocmd VimEnter * silent! call LspOptionsSet(lspOpts)
augroup END
```

**Comment by maximyurevich** [comment]
> > Hello @andlrc, multiple servers starting working only after `LspServer restart`. Is this normal behavior for now?
> 
> No, how have you configured your servers?

rootSearch works too only works on LspServer restart, maybe it's a vim-plug problem

**Comment by andlrc** [comment]
@maximyurevich I would assume with a setup like this:

> ```viml
> 
> let g:lspServers = [
> 	\  {
> 	\	    'name': 'typescript',
> 	\	    'filetype': ['javascript', 'typescript', 'javascriptreact', 'typescriptreact'],
> 	\	    'path': mason_dir.'typescript-language-server',
> 	\	    'args': ['--stdio'],
>         \     'rootSearch': ['package.json']
> 	\   },
> 	\   {
> 	\	    'name': 'deno',
> 	\	    'filetype': ['javascript', 'typescript', 'javascriptreact', 'typescriptreact'],
> 	\	    'path': 'deno',
> 	\	    'args': ['lsp'],
>         \     'initializationOptions': {
>         \       'enable': v:true,
>         \       'lint': v:true
>         \     },
>         \     'rootSearch': ['deno.json', 'deno.jsonc']
> 	\   },
> ...
> 	\   {
> 	\	    'name': 'efm-langserver',
> 	\	    'filetype': [
>         \                   'vim',
>         \                   'javascript', 
>         \                   'typescript', 
>         \                   'javascriptreact', 
>         \                   'typescriptreact',
>         \                   'python',
>         \                   'ruby'
>         \                 ],
> 	\	    'path': 'efm-langserver',
> 	\	    'args': [],
>         \   },
> 	\ ]
> ```

Then only one server would run, as servers with unmatched `rootSearch` are scored below servers without `rootSearch`, and are therefore skipped.

While servers with a matched `rootSearch` will score above servers without a `rootSearch`, causing the once without `rootSearch` to be skipped.

Can you try adding ```
'rootSearch': ['package.json'] + ['deno.json', 'deno.jsonc']
```

To the `efm-langserver` as I think it will then run together with either `deno` or `typescript-language-server`.

I don't understand why `LspRestart` should do anything, as it only restarts already "used" servers.

**Comment by maximyurevich** [comment]
> @maximyurevich I would assume with a setup like this:
> 
> > ```viml
> > let g:lspServers = [
> > 	\  {
> > 	\	    'name': 'typescript',
> > 	\	    'filetype': ['javascript', 'typescript', 'javascriptreact', 'typescriptreact'],
> > 	\	    'path': mason_dir.'typescript-language-server',
> > 	\	    'args': ['--stdio'],
> >         \     'rootSearch': ['package.json']
> > 	\   },
> > 	\   {
> > 	\	    'name': 'deno',
> > 	\	    'filetype': ['javascript', 'typescript', 'javascriptreact', 'typescriptreact'],
> > 	\	    'path': 'deno',
> > 	\	    'args': ['lsp'],
> >         \     'initializationOptions': {
> >         \       'enable': v:true,
> >         \       'lint': v:true
> >         \     },
> >         \     'rootSearch': ['deno.json', 'deno.jsonc']
> > 	\   },
> > ...
> > 	\   {
> > 	\	    'name': 'efm-langserver',
> > 	\	    'filetype': [
> >         \                   'vim',
> >         \                   'javascript', 
> >         \                   'typescript', 
> >         \                   'javascriptreact', 
> >         \                   'typescriptreact',
> >         \                   'python',
> >         \                   'ruby'
> >         \                 ],
> > 	\	    'path': 'efm-langserver',
> > 	\	    'args': [],
> >         \   },
> > 	\ ]
> > ```
> 
> Then only one server would run, as servers with unmatched `rootSearch` are scored below servers without `rootSearch`, and are therefore skipped.
> 
> While servers with a matched `rootSearch` will score above servers without a `rootSearch`, causing the once without `rootSearch` to be skipped.
> 
> Can you try adding
> 
> ```
> 'rootSearch': ['package.json'] + ['deno.json', 'deno.jsonc']
> ```
> 
> To the `efm-langserver` as I think it will then run together with either `deno` or `typescript-language-server`.
> 
> I don't understand why `LspRestart` should do anything, as it only restarts already "used" servers.

What plugin manager you use?

**Comment by andlrc** [comment]
> What plugin manager you use?

`:h packadd`

I made https://github.com/yegappan/lsp/pull/261 as an attempt to separate the "should a server be started" concern into a new configuration option. Can you take a look and try it out?

**Comment by maximyurevich** [comment]
> I made #261 as an attempt to separate the "should a server be started" concern into a new configuration option. Can you take a look and try it out?

I replaced rootSearch with runIfSearch and the servers start correctly according to the given patterns, except efm

**Comment by andlrc** [comment]
> I replaced rootSearch with runIfSearch and the servers start correctly according to the given patterns, except efm

If you add `'debug': true` to `efm-langserver` then you'll get logging, you can run `:LspServer debug errors` or `:LspServer debug messages` and splits will open with error or logging files.

Pay attention to the error file for `efm-langserver`.

**Comment by maximyurevich** [comment]
> > I replaced rootSearch with runIfSearch and the servers start correctly according to the given patterns, except efm
> 
> If you add `'debug': true` to `efm-langserver` then you'll get logging, you can run `:LspServer debug errors` or `:LspServer debug messages` and splits will open with error or logging files.
> 
> Pay attention to the error file for `efm-langserver`. And possible also the log file for `efm-langserver`.

Where is it better to put efm, at the start or at the end?

**Comment by andlrc** [comment]
> Where is it better to put efm, at the start or at the end? When on start, efm works, but not lsp, else lsp works, but not efm

I don't think you checked out the commits at #261 . If you want to override behaviors of a given language server, then see `:h lsp-features`.

**Comment by maximyurevich** [comment]
> :h lsp-features
https://i.imgur.com/vK0an42.png

**Comment by andlrc** [comment]
> > :h lsp-features
> 
> https://i.imgur.com/vK0an42.png

You need to first generate helptags, see `:h :helptags`, basically then `:helptags ALL` should work for you.

**Comment by maximyurevich** [comment]
I can't get efm log without LspServer restart
[lsp-typescript.log](https://github.com/yegappan/lsp/files/11239755/lsp-typescript.log)
[lsp-efm-langserver.log](https://github.com/yegappan/lsp/files/11239756/lsp-efm-langserver.log)

**Comment by andlrc** [comment]
> I can't get efm log without LspServer restart [lsp-typescript.log](https://github.com/yegappan/lsp/files/11239755/lsp-typescript.log) [lsp-efm-langserver.log](https://github.com/yegappan/lsp/files/11239756/lsp-efm-langserver.log)

What about the `.err` files?

**Comment by maximyurevich** [comment]
> > I can't get efm log without LspServer restart [lsp-typescript.log](https://github.com/yegappan/lsp/files/11239755/lsp-typescript.log) [lsp-efm-langserver.log](https://github.com/yegappan/lsp/files/11239756/lsp-efm-langserver.log)
> 
> What about the `.err` files?

https://imgur.com/7TgihlT.png

**Comment by maximyurevich** [comment]
> > It looks like you are trying to use `efm-langserver` for formatting, of so then remember to specify `features: { documentFormatting: true } }` see `:h lsp-features`. And possible also `initializationOptions: {documentFormatting: true }}` see https://github.com/mattn/efm-langserver#configuration
> 
> @maximyurevich ^

I want to use efm for linting and formatting

**Comment by maximyurevich** [comment]
> > It looks like you are trying to use `efm-langserver` for formatting, of so then remember to specify `features: { documentFormatting: true } }` see `:h lsp-features`. And possible also `initializationOptions: {documentFormatting: true }}` see https://github.com/mattn/efm-langserver#configuration
> 
> @maximyurevich ^

thank you :) now all works

**Comment by andlrc** [comment]
Happy it works. And thanks for confirming that multiple language servers works properly. I think that it might be time to close this issue then.

**Comment by maximyurevich** [comment]
> Happy it works. And thanks for confirming that multiple language servers works properly. I think that it might be time to close this issue then.

Yes
